### PR TITLE
feat: faster and more verbose VK generation

### PR DESCRIPTION
**Pull request by mm-zk**
# What ❔

* allows verification keys to be generated in parallel
* adds support for progress bar
* also updated verification keys for the recent circuit changes.

 
## Why ❔

* Verification key generation can take over 30 minutes, with ability to generate them in parallel (and enough ram) - this can be reduced multiple times.
* Progress bar hooks allows more information for the end user.

Part of EVM-642 

Example how it looks like with progress bar:
image

### CODE REVIEW

File: src/compute_setups.rs [hunk 1]
```diff
@@ -412,76 +414,155 @@ pub fn generate_circuit_setup_data(
...
 
+fn generate_vk_and_finalization_hint(
+    circuit: ZkSyncBaseLayerCircuit,
+    worker: &Worker,
+) -> (
+    ZkSyncBaseLayerVerificationKey,
+    ZkSyncBaseLayerFinalizationHint,
+) {
+    let circuit_type = circuit.numeric_circuit_type();
+    //let worker = Worker::new_with_num_threads(threads_per_worker);
```
Comment (0xVolosnikov) [inline src/compute_setups.rs]: Cleanup

File: src/compute_setups.rs [hunk 2]
```diff
@@ -412,76 +414,155 @@ pub fn generate_circuit_setup_data(
...
 /// For backwards compatibility (as zksync-era uses this method).
 /// For new cases please use generate_recursive_layer_vks directly.
 pub fn generate_recursive_layer_vks_and_proofs(
     source: &mut dyn SetupDataSource,
 ) -> crate::data_source::SourceResult<()> {
-    generate_recursive_layer_vks(source)
+    generate_recursive_layer_vks(source, None, || {})
+}
+
+fn generate_vk_and_finalization_hint_for_recursion(
```
Comment (0xVolosnikov) [inline src/compute_setups.rs:493]: Wouldn't it be cleaner using generic for recursion and base versions?

```rust 
fn generate_vk_and_finalization_hint<C, K, H>(circuit: C, worker: &Worker) -> (K, H) {
...
}
```

**Review by 0xVolosnikov — APPROVED** [review]

File: src/compute_setups.rs (hunk 1 above)
Comment (mm-zk) [reply src/compute_setups.rs]: done. thx.

File: src/compute_setups.rs (hunk 2 above)
Comment (mm-zk) [reply src/compute_setups.rs:493]: that's a nice idea - but it requires a little bit more changes in other methods too. I'll do it in a separate PR.

**Pull request merged**